docs/guides/end-to-end-testing/testing-strategies/azure-active-directory-authentication.mdx

```diff
@@ -0,0 +1,242 @@
+    },
+    ({ password }) => {
+      cy.get('input[type="password"]').type(password, {
+        log: false,
+      })
+      cy.get('input[type="submit"]').click()
+      cy.get('#idBtn_Back').click()
+    }
+  )
+
+  // Ensure Microsoft has redirected us back to the sample app with our logged in user.
+  cy.url().should('equal', 'http://localhost:3000/')
+  cy.get('#welcome-div').should(
+    'contain',
+    `Welcome ${Cypress.env('aad_username')}!`
+  )
+}
+
+Cypress.Commands.add('loginToAAD', (username: string, password: string) => {
+  const log = Cypress.log({
+    displayName: 'Azure Active Directory Login',
+    message: [`🔐 Authenticating | ${username}`],
+    // @ts-ignore
+    autoEnd: false,
+  })
+  log.snapshot('before')
+
+  loginViaAAD(username, password)
+
+  log.snapshot('after')
+  log.end()
+})
+```
+
+Now, we can use our `loginToAAD` command in the test. Below is our test to login
+as a user via Azure Active Directory and run a basic sanity check.
+
+```js
+describe('Azure Active Directory Authentication', () => {
+  beforeEach(() => {
+    // log into Azure Active Directory through our sample SPA using our custom command
+    cy.loginToAAD(Cypress.env('aad_username'), Cypress.env('aad_password'))
+    cy.visit('http://localhost:3000')
+  })
+
+  it('verifies the user logged in has the correct name', () => {
+    cy.get('#table-body-div td:contains("name") + td').should(
+      'contain',
+      `${Cypress.env('aad_name')}`
+    )
+  })
+
+  it('verifies the user logged in has the correct preferred name', () => {
+    cy.get('#table-body-div td:contains("preferred_username") + td').should(
+      'contain',
+      `${Cypress.env('aad_username')}`
+    )
+  })
+})
+```
+
+<DocsVideo src="https://vimeo.com/789095126"></DocsVideo>
+
+We now have working authentication and tests! But we are logging in before every
+test, which is not only time consuming, but also can lead to API rate limiting
+due to the number of requests.
+
+For this, we can refactor our login command to take advantage of
+[`cy.session()`](/api/commands/session) to store our logged in user so we don't
+have to reauthenticate before every test.
+
+```js
+Cypress.Commands.add('loginToAAD', (username: string, password: string) => {
+  cy.session(
+    `aad-${username}`,
+    () => {
+      const log = Cypress.log({
+        displayName: 'Azure Active Directory Login',
+        message: [`🔐 Authenticating | ${username}`],
+        // @ts-ignore
+        autoEnd: false,
+      })
+
+      log.snapshot('before')
+
+      loginViaAAD(username, password)
+
+      log.snapshot('after')
+      log.end()
+    },
+    {
+      validate: () => {
+        // this is a very basic form of session validation for this demo.
+        // depending on your needs, something more verbose might be needed
+        cy.visit('http://localhost:3000')
+        cy.get('#welcome-div').should(
+          'contain',
+          `Welcome ${Cypress.env('aad_username')}!`
+        )
+      },
```

Comment:
Not a lot of great options for Oauth it looks like. I looked at some of the values in cookies, local, and session storage, but couldnt find anything deterministic for what would guarantee the user is logged it. An approach I thought about using here was using the AAD library to figure out who is logged into the session and verifying the user that way, but I figured that might be too much for the simple guide. What do you think? Should we opt for that approach to avoid the visit?

Reply:
It's hard to say - this type of validation is "heavier" and "slower" for users but might be "safest" for the longevity of the guide